src/ripple/app/tx/impl/details/NFTokenUtils.cpp

```diff
@@ -636,5 +674,122 @@ deleteTokenOffer(ApplyView& view, std::shared_ptr<SLE> const& offer)
     return true;
 }
 
+bool
+repairNFTokenDirectoryLinks(ApplyView& view, AccountID const& owner)
+{
+    bool didRepair = false;
+
+    auto const last = keylet::nftpage_max(owner);
+
+    std::shared_ptr<SLE> page = view.peek(Keylet(
+        ltNFTOKEN_PAGE,
+        view.succ(keylet::nftpage_min(owner).key, last.key.next())
+            .value_or(last.key)));
+
+    if (!page)
+        return didRepair;
+
+    if (page->key() == last.key)
+    {
+        // There's only one page in this entire directory.  There should be
+        // no links on that page.
+        bool const nextPresent = page->isFieldPresent(sfNextPageMin);
+        bool const prevPresent = page->isFieldPresent(sfPreviousPageMin);
+        if (nextPresent || prevPresent)
+        {
+            didRepair = true;
+            if (prevPresent)
+                page->makeFieldAbsent(sfPreviousPageMin);
+            if (nextPresent)
+                page->makeFieldAbsent(sfNextPageMin);
+            view.update(page);
+        }
+        return didRepair;
+    }
+
+    // First page is not the same as last page.  The first page should not
+    // contain a previous link.
+    if (page->isFieldPresent(sfPreviousPageMin))
+    {
+        didRepair = true;
+        page->makeFieldAbsent(sfPreviousPageMin);
+        view.update(page);
+    }
+
+    std::shared_ptr<SLE> nextPage;
+    while (
```

Comment:
would this while-loop require performance testing?

Reply:
When the owner has a `last` page, is that sufficient to deduce that this particular owner is not a "victim" of the broken link?  Is there still the need to execute this loop? I'm just wondering if that's a viable approach instead of iterating through the entire doubly linked `NFTokenPage` structure every time.

Reply:
Good question.  But no, the presence of a valid `last` page is not sufficient to know that the directory tree is undamaged.  In the two current on-ledger cases both of the directories have valid `last` pages and there is a single break between two pages in the middle of the directory.

The way this happens is that

1. The last page is removed.  This leaves the next page down as the new `last` page.  But that new `last` page does not have the official `last` index for this directory.
2. The directory owner adds an NFToken with an ID greater the index of the current `last` page.  This causes a new `last` page to be created with the correct `last` index.  But the new `last` page does not have a link to the previous page.  So the link is missing in the middle of the chain.

It's possible to have the case where the last page is simply missing.  And I also have a unit test where I repair that situation.  But what I've seen is that, typically, the link is missing in the middle of the chain.

It would be possible for there to be more than one broken link in the chain.  But I've not seen that in the wild.

Does that make sense?  Questions?

Reply:
Thanks for the detailed explanation! This makes a lot of sense!